### ISSUE TITLE
HLIL sometimes do not simplify array access

### ISSUE THREAD

**Issue by xusheng6** [Vector35/binaryninja-api#4251]
![Screenshot 2023-04-20 at 3 59 26 PM](https://user-images.githubusercontent.com/94503187/233299815-5d2dadfc-635d-4c3f-9089-b1a85c79eeb3.png)

In the screenshot above, we can see that the HLIL sometimes correctly recognizes the array access and sometimes do not. Steps to reproduce:

1. Open the binary [once_and_for_all.zip](https://github.com/Vector35/binaryninja-api/files/11283783/once_and_for_all.zip)
2. Go to types view, create the structure:
```
struct weapon __packed
{
    int64_t size;
    int64_t data;
};
```
3. Go to 0x202080, and define a data variable as `weapon[0xf]`
4. View the code at 0xcd4

I notice all of the failed cases are within the `if` statement. There might be a bug in the HLIL simplifier that skips the simplification of the expression

**Comment by xusheng6** [comment]
Not to be confused with https://github.com/Vector35/binaryninja-api/issues/2570, the latter one mainly requests array access to detected automatically. While in this issue, the array access is known, but the HLIL fails to display it in the optimal way

**Comment by lwerdna** [comment]
This is a minimized repro for one of the situations: [hello.zip](https://github.com/Vector35/binaryninja-api/files/13357029/hello.zip)

Open, go to `100004000: array` and set type to `uint32_t array[0x10]`. Then look at `_main()`:

![image](https://github.com/Vector35/binaryninja-api/assets/474191/a70e2912-ca32-435d-af50-91b9d6b40ab2)

**Comment by mostobriv** [comment]
May it be related to HLIL "if" optimisation issue (#4223)? Look like HLIL get hard times when optimising "if" conditions.

**Comment by D0ntPanic** [comment]
Fixed in 4.1.5124
